plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,168 @@
+package whatap
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/outputs"
+
+	whatap_io "github.com/whatap/go-api/common/io"
+	whatap_pack "github.com/whatap/go-api/common/lang/pack"
+	whatap_hash "github.com/whatap/go-api/common/util/hash"
+)
+
+const (
+	NetSrcAgentOneway = 10
+)
+
+type Whatap struct {
+	License string          `toml:"license"`
+	Servers []string        `toml:"servers"`
+	Pcode   int64           `toml:"pcode"`
+	Timeout config.Duration `toml:"timeout"`
+	Log     telegraf.Logger `toml:"-"`
+	oname   string
+	oid     int32
+	conn    net.Conn
+	dest    int
+}
+
+var sampleConfig = `
+  ## You can create a project on the WhaTap site(https://www.whatap.io) 
+  ## to get license, project code and server IP information.
+
+  ## WhaTap license. Required
+  license = "xxxx-xxxx-xxxx"
+
+  ## WhaTap project code. Required
+  pcode = 1111
+
+  ## WhaTap server IP. Required
+  ## Put multiple IPs. ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+  servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	w.dest++
+	if w.dest >= len(w.Servers) {
+		w.dest = 0
+	}
+
+	u, err := url.Parse(w.Servers[w.dest])
+	if err != nil {
+		return fmt.Errorf("invalid address: %s", w.Servers[w.dest])
+	}
+	if u.Scheme != "tcp" {
+		return fmt.Errorf("only tcp is supported: %s", w.Servers[w.dest])
+	}
+	client, err := net.DialTimeout(u.Scheme, u.Host, time.Duration(w.Timeout))
+	if err != nil {
+		return err
+	}
+
+	w.conn = client.(*net.TCPConn)
+	w.Log.Info("Connected ", u.String())
+	return nil
+}
+
+func (w *Whatap) Close() error {
+	if w.conn == nil {
+		return nil
+	}
+	err := w.conn.Close()
+	w.conn = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Plugin to send metrics to a WhaTap server"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.conn == nil {
+		if err := w.Connect(); err != nil {
+			_ = w.Close()
```

Comment:
Does this really make sense to close the connection if connecting failed?

Reply:
Deleted that phrase.